xnmt/transformer.py

```diff
@@ -0,0 +1,321 @@
+
+  def __call__(self, x, z=None, mask=None):
+    h = self.h
+    if self.is_self_attention:
+      Q = self.W_Q(x)
+      K = self.W_K(x)
+      V = self.W_V(x)
+    else:
+      Q = self.W_Q(x)
+      K = self.W_K(z)
+      V = self.W_V(z)
+
+    (n_units, n_querys), batch = Q.dim()
+    (_, n_keys), _ = K.dim()
+
+    # Calculate Attention Scores with Mask for Zero-padded Areas
+    # Perform Multi-head Attention using pseudo batching
+    # all together at once for efficiency
+
+    batch_Q = dy.concatenate_to_batch(split_rows(Q, h))
+    batch_K = dy.concatenate_to_batch(split_rows(K, h))
+    batch_V = dy.concatenate_to_batch(split_rows(V, h))
+
+    assert(batch_Q.dim() == (n_units // h, n_querys), batch * h)
+    assert(batch_K.dim() == (n_units // h, n_keys), batch * h)
+    assert(batch_V.dim() == (n_units // h, n_keys), batch * h)
+
+    mask = np.concatenate([mask] * h, axis=0)
+    mask = dy.inputTensor(np.moveaxis(mask, [1, 0, 2], [0, 2, 1]), batched=True)
+    batch_A = (dy.transpose(batch_Q) * batch_K) * self.scale_score
+    batch_A = dy.cmult(batch_A, mask) + (1 - mask)*MIN_VALUE
+
+    sent_len = batch_A.dim()[0][0]
+    if sent_len == 1:
+        batch_A = dy.softmax(batch_A)
+    else:
+        batch_A = dy.softmax(batch_A, d=1)
+
+    # TODO: Check whether this is correct after masking
+    batch_A = dy.cmult(batch_A, mask)
+    assert (batch_A.dim() == ((n_querys, n_keys), batch * h))
+
+    # TODO: Check if attention dropout needs to be applied here
+    batch_C = dy.transpose(batch_A * dy.transpose(batch_V))
+
+    # batch_C = batch_V * dy.transpose(batch_A)  # TODO: Check the correctness of this step
+    assert (batch_C.dim() == ((n_units // h, n_querys), batch * h))
+
+    C = dy.concatenate(split_batch(batch_C, h), d=0)
+    assert (C.dim() == ((n_units, n_querys), batch))
+    C = self.finishing_linear_layer(C)
+    return C
+
+
+class FeedForwardLayer(object):
+  def __init__(self, dy_model, n_units):
+    n_inner_units = n_units * 4
+    self.W_1 = Linear(dy_model, n_units, n_inner_units)
+    self.W_2 = Linear(dy_model, n_inner_units, n_units)
+    self.act = dy.rectify
+
+  def __call__(self, e):
+    e = self.W_1(e, reconstruct_shape=False, timedistributed=True)
+    e = self.act(e)
+    e = self.W_2(e, reconstruct_shape=False, timedistributed=True)
+    return e
+
+
+class EncoderLayer():
+  def __init__(self, dy_model, n_units, h=1):
+    self.self_attention = MultiHeadAttention(dy_model, n_units, h)
+    self.feed_forward = FeedForwardLayer(dy_model, n_units)
+    self.ln_1 = LayerNorm(dy_model, n_units)
+    self.ln_2 = LayerNorm(dy_model, n_units)
+
+  def set_dropout(self, dropout):
+    self.dropout = dropout
+
+  def __call__(self, e, xx_mask):
+    self.self_attention.set_dropout(self.dropout)
+    sub = self.self_attention(e, e, xx_mask)
+    e = e + dy.dropout(sub, self.dropout)
+    e = self.ln_1(e)
+
+    sub = self.feed_forward(e)
+    e = e + dy.dropout(sub, self.dropout)
```

Comment:
One more thing: it would be better to add an if statement and skip the dropout operation for self.dropout==0.0

Reply:
This is a good idea. I will do it in my future commits.

Reply:
Yes, please do.